django/db/models/expressions.py

```diff
@@ -394,14 +395,26 @@ def as_sql(self, compiler, connection):
             rhs_output = self.rhs.output_field
         except FieldError:
             rhs_output = None
+
+        # TODO: replace this with CombinedExpression.dispatch
         if (not connection.features.has_native_duration_field and
                 ((lhs_output and lhs_output.get_internal_type() == 'DurationField')
                 or (rhs_output and rhs_output.get_internal_type() == 'DurationField'))):
             return DurationExpression(self.lhs, self.connector, self.rhs).as_sql(compiler, connection)
-        if (lhs_output and rhs_output and self.connector == self.SUB and
-            lhs_output.get_internal_type() in {'DateField', 'DateTimeField', 'TimeField'} and
-                lhs_output.get_internal_type() == lhs_output.get_internal_type()):
-            return TemporalSubtraction(self.lhs, self.rhs).as_sql(compiler, connection)
+
+        if lhs_output or rhs_output:
+            dispatcher = self.connector_dispatchers.get(self.connector)
+            try:
+                combined_expression = dispatcher(lhs_output, rhs_output) if dispatcher else CombinedExpression
```

Comment:
The fallback is meant for unregistered operators.

Speaking of which: I think I might be able to get rid of `connection.ops.combine_expression` call in favor of dispatching expression suitable for the particular database backend. This will be a bit more invasive, but I think also cleaner in the long run. What do you think?